### ISSUE TITLE
Defined Tags with interpolated values referencing not-yet-created resources cause Terraform apply to fail

### ISSUE THREAD

**Issue by jearls** [terraform-providers/terraform-provider-oci#911]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version and Provider Version

Terraform version `0.11.14`
Provider version `3.49.0`

### Affected Resource(s)

I saw this in `oci_core_instance` but assume that it affects any resource that can use defined tags.

### Terraform Configuration Files

```hcl
locals {
  compartment_id = "…"
  availability_domain = "…"
}

resource "oci_identity_tag_namespace" "testing" {
    provider       = "oci"
    compartment_id = "${local.compartment_id}"
    description    = "testing"
    name           = "testing"
}

resource "oci_identity_tag" "testing_foo" {
    provider         = "oci"
    tag_namespace_id = "${oci_identity_tag_namespace.testing.id}"
    description      = "foo"
    name             = "foo"
}

resource "oci_core_instance" "test1" {
    provider       = "oci"
    count = 1
    availability_domain = "${local.availability_domain}"
    compartment_id      = "${local.compartment_id}"
    shape               = "VM.Standard2.1"
    display_name   = "test1"
    source_details {
        source_type             = "image"
        source_id               = "ocid1.image.oc1.phx.aaaaaaaadtmpmfm77czi5ghi5zh7uvkguu6dsecsg7kuo3eigc5663und4za" // Oracle-Linux-7.7-2019.08.28-0
        boot_volume_size_in_gbs = 60
    }
    preserve_boot_volume = "false"
}

resource "oci_core_instance" "test2" {
    provider       = "oci"
    count = 1
    availability_domain = "${local.availability_domain}"
    compartment_id      = "${local.compartment_id}"
    shape               = "VM.Standard2.1"
    display_name   = "test1"
    source_details {
        source_type             = "image"
        source_id               = "ocid1.image.oc1.phx.aaaaaaaadtmpmfm77czi5ghi5zh7uvkguu6dsecsg7kuo3eigc5663und4za" // Oracle-Linux-7.7-2019.08.28-0
        boot_volume_size_in_gbs = 60
    }
    preserve_boot_volume = "false"
    defined_tags {
        testing.foo = "${oci_core_instance.test1.id}"
    }
    depends_on  = [ "oci_identity_tag_namespace.testing", "oci_identity_tag.testing_foo", "oci_core_instance.test1" ]
}
```

### Debug Output

https://gist.github.com/jearls/794f412169c25416e647a0eedc5a3dd8

### Expected Behavior

The instance "test2" should be created with a defined tag "testing.foo" equal to the OCID of the instance "test".

### Actual Behavior

1. `terraform plan` does not show defined tags being generated on the `test2` instance
2. `terraform apply` fails with the error `oci_core_instance.test2: diffs didn't match during apply. This is a bug with Terraform and should be reported as a GitHub Issue.`

### Steps to Reproduce

0. See configuration above.
1. `terraform apply`

**Comment by jearls** [comment]
Note - this apparently works in terraform `0.12` (although i'm finding a different issue in tf `0.12` regarding defined tags which i'll submit in a different issue)

**Comment by parrneet** [comment]
Hi @jearls 

I tried a few things like using ${map()} and lookup via datasources and some other workarounds suggested [here](https://github.com/hashicorp/terraform/issues/444) but could not interpolate within maps
Interpolation inside variables and maps has been a bug with Terraform and they have fixed this in v0.12. I would suggest upgrading 

Closing the issue, please reopen if you have any questions.

**Comment by jearls** [comment]
@parrneet this is not a general issue with interpolating.  it's the specific case of an interpolation string used within defined tags that refers to a resource that has not yet been created.  Once terraform creates the resource, you can re-run the terraform apply and it will properly interpolate the values.

Upgrading to 0.12 is not an option we can undertake lightly.  that involves many changes throughout our automated templating system and will take significant time to test and ensure it does not bfreak anything.
